File: solutions/csharp/sieve/1/Sieve.cs
using System;
using System.Collections.Generic;

public static class Sieve
{
    public static int[] Primes(int limit)
    {
        if (limit < 0)
            throw new ArgumentOutOfRangeException($"No negaitive numbers for {nameof(limit)}");

        HashSet<int> visited = new();
        List<int> primes = new();
        for (int num = 2; num <= limit; num++)
        {
            if (!visited.Contains(num))
            {
                primes.Add(num);
                for (int factor = num; factor <= limit; factor += num)
                {
                    visited.Add(factor);
                }
            }
        }

        return primes.ToArray();
    }
}

File: solutions/cpp/sieve/1/sieve.cpp
#include "sieve.h"
#include <vector>

namespace sieve {
std::vector<int> primes(int num) {
  if (num < 2) {
    return std::vector<int>{};
  }
  std::vector<int> prime_list{2};
  for (int candidate = prime_list.back() + 1; candidate <= num; ++candidate) {
    bool is_prime = true;
    for (int i = 0; i < static_cast<int>(prime_list.size()) && is_prime; ++i) {
      is_prime = candidate % prime_list[i] != 0;
    }
    if (is_prime) {
      prime_list.push_back(candidate);
    }
  }

  return prime_list;
}
} // namespace sieve
